File: windows/BlueLink.App/Session/TransferSourceIdentity.cs
using BlueLink.Domain;

namespace BlueLink.Session;

internal static class TransferSourceIdentity
{
    public static void Validate(TransferItem previous, long size, ReadOnlySpan<byte> sha256)
    {
        if (previous.SourceSha256 is not { Length: 64 } expected || !expected.All(Uri.IsHexDigit))
            throw new InvalidOperationException(Localization.Strings.Get("旧任务缺少文件指纹，请重新选择文件发送"));
        if (size != previous.TotalBytes || !expected.Equals(Convert.ToHexString(sha256), StringComparison.OrdinalIgnoreCase))
            throw new InvalidOperationException(Localization.Strings.Get("文件内容已变化，请重新选择文件发送"));
    }
}

File: android/app/src/main/java/com/bluelink/android/session/TransferSourceIdentity.kt
package com.bluelink.android.session

import com.bluelink.android.domain.TransferItem

/** Compare the immutable retry snapshot before publishing an offer or resuming a checkpoint. */
internal object TransferSourceIdentity {
    enum class Problem { MISSING, CHANGED }
    fun problem(previous: TransferItem, size: Long, sha256: ByteArray): Problem? {
        val expected = previous.sourceSha256?.takeIf { it.matches(Regex("[0-9a-fA-F]{64}")) }
            ?: return Problem.MISSING
        val actual = sha256.joinToString("") { "%02x".format(it) }
        return if (size != previous.totalBytes || !expected.equals(actual, ignoreCase = true)) Problem.CHANGED else null
    }
}
